errors: eliminate all overhead for hidden calls

mmomtchev opened the pull request:
##### Checklist
- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

<!--
Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

This in response to #35386

puzpuzpuz commented on lib/internal/errors.js:
```diff
@@ -393,6 +378,16 @@ function uvErrmapGet(name) {
 
+function captureLargerStackTrace(err) {
+  userStackTraceLimit = Error.stackTraceLimit;
+  Error.stackTraceLimit = 20;
```
Why `20` is chosen as the trace limit? I don't see such restriction in the old code base.

mmomtchev replied:
It is an arbitrary limit, yes
There is no easy way to determine in advance how many additional calls will be there
Unless you successively try larger and larger sizes

puzpuzpuz replied:
If there is no good reason behind the constant, I'd suggest to remove the limit here. Otherwise, this PR may introduce a regression in user applications' exception handling. We're not in control of the user code and it may have arbitrary call depth and/or `Error.stackTraceLimit` value.

puzpuzpuz commented on lib/internal/errors.js:
```diff
@@ -64,6 +65,16 @@ const prepareStackTrace = (globalThis, error, trace) => {
     return f(error, trace);
   }
 
+  for (let l = trace.length - 1; l >= 0; l--) {
```
This PR seems to move the "overhead" from success-path to exception-path. It might be a good idea to benchmark both paths, say, for `buf.write()` method, just to be able to understand whether there is any improvement/regression.

mmomtchev replied:
misc/hidestackframes.js n=100000 type="hide-err": 22,456.052314327284
misc/hidestackframes.js n=100000 type="direct-err": 20,387.649071345215
misc/hidestackframes.js n=100000 type="hide-noerr": 10,342,015.627819816
misc/hidestackframes.js n=100000 type="direct-noerr": 10,862,952.985139482


```js
'use strict';

const common = require('../common.js');

const bench = common.createBenchmark(main, {
  type: ['hide-err', 'direct-err', 'hide-noerr', 'direct-noerr'],
  n: [100000]
}, {
  flags: ['--expose-internals']
});

function main({ n, type }) {
  const {
    hideStackFrames,
    codes: {
      ERR_INVALID_ARG_TYPE,
    },
  } = require('internal/errors');

  const testfn = (value) => {
    if (typeof value !== 'number') {
      throw new ERR_INVALID_ARG_TYPE('Benchmark', 'number', value);
    }
    let r = 1;
    for (let i = 0; i < 4; i++)
      r *= (i / 2);
    return r;
  };

  let fn = testfn;
  if (type.startsWith('hide'))
    fn = hideStackFrames(testfn);
  let value = 42;
  if (type.endsWith('-err'))
    value = 'err';

  bench.start();

  for (let i = 0; i < n; i++) {
    try {
      fn(value);
    } catch (e) {
      // No-op
    }
  }

  bench.end(n);
}
```

mmomtchev replied:
The error path is already far too slow for the regression to be noticeable, there is a factor of 500

mmomtchev replied:
Without the PR:
misc/hidestackframes.js n=100000 type="hide-err": 24,434.469081672654
misc/hidestackframes.js n=100000 type="direct-err": 24,511.501403118
misc/hidestackframes.js n=100000 type="hide-noerr": 6,504,481.750538165
misc/hidestackframes.js n=100000 type="direct-noerr": 10,724,409.173530914

mmomtchev replied:
I have added a small arithmetic loop to your test so that it is more realistic, otherwise the difference is huge but unrealistic

mmomtchev replied:
This is without the arithmetic operations:
```
mmom@mmom-workstation:~/src/node$ ./node benchmark/misc/hidestackframes.js  # without PR
misc/hidestackframes.js n=100000 type="hide-err": 24,542.416148820295
misc/hidestackframes.js n=100000 type="direct-err": 24,689.453077934282
misc/hidestackframes.js n=100000 type="hide-noerr": 7,136,211.80048265
misc/hidestackframes.js n=100000 type="direct-noerr": 10,672,188.884082846
mmom@mmom-workstation:~/src/node$ ./node benchmark/misc/hidestackframes.js    # with PR
misc/hidestackframes.js n=100000 type="hide-err": 22,586.849942378212
misc/hidestackframes.js n=100000 type="direct-err": 20,270.969751583732
misc/hidestackframes.js n=100000 type="hide-noerr": 11,406,198.333485987
misc/hidestackframes.js n=100000 type="direct-noerr": 11,542,427.250100967
mmom@mmom-workstation:~/src/node$ 
```

puzpuzpuz replied:
It may be a good idea to place the benchmark into `benchmark/misc`. Then you could use compare script to compare new vs old performance. See:
https://github.com/nodejs/node/blob/master/doc/guides/writing-and-running-benchmarks.md#comparing-nodejs-versions

Trott replied:
Here's a benchmark run in CI. https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/668/

puzpuzpuz commented on lib/internal/errors.js:
```diff
@@ -393,6 +378,16 @@ function uvErrmapGet(name) {
   return uvBinding.errmap.get(name);
 }
 
+function captureLargerStackTrace(err) {
```
Do we still need this function? It looks like it could be replaced with `Error.captureStackTrace(ex)`.

mmomtchev replied:
I still need to extend the stack trace size before capturing it. The way the code works is that the stack trace size is set to 0 at the ```Error``` creation (so no stack trace is collected), then it manually invokes ```Error.captureStackTrace(ex)```. That ```captureLargerStackTrace(err)``` is there mainly to centralize at one place extending the stack size and then restoring it back.

puzpuzpuz replied:
All invocations of `captureLargerStackTrace()` are made when `Error.stackTraceLimit` is already restored from `0` to the original value, so I still don't get why it's needed. Am I missing something?

mmomtchev replied:
I need to increase it from the original value to Infinity, since I will be removing some elements and I don't know their number in advance. Then I need to cut off the stack trace to the original value length, so that I can return the exact number of stack trace elements that the user asked for - usually 10.

puzpuzpuz replied:
I see now. Then this approach with `Infinity` may lead to increased memory consumption in user applications, especially when they make standard library calls in deeply nested and/or recursive functions.

It can be demonstrated on the following example:
```js
// comment to avoid crashing
Error.stackTraceLimit = Infinity;
function foo(i) {
    if (i === 32000) {
        throw new Error('bar');
    }
    return foo(i + 1);
}
foo(1);
```

If you run it with `node --max-old-space-size=4 --stack-size=32000 recursive-error.js`, it'll crash. Without the `Error.stackTraceLimit` override, the script stops on unhandled error as expected.

Thus, the current implementation seems to be introducing a potential regressions. We could change `Infinity` to `userStackTraceLimit + N` where `N` is the max depth of standard library validation function calls, but that's a bad approach. Do you have an idea for another workaround here?

mmomtchev replied:
I have another idea, but it is even uglier - rename all functions that are to be hidden to the same name and then use the V8 built-in `Error.captureStackTrace` feature - which is much faster
But it is going to make life even harder for internals developers

mmomtchev replied:
In fact, I think that aside from some very edge artificial cases - where the heap is smaller than the stack, we should be ok - if you have a call stack that is so large, that it doesn't fit into the heap - you are bound to run into problems anyway

puzpuzpuz replied:
Heap may be large enough to fit the call stack, yet it may not have enough free space to allocate an exception as it also contains other JS objects. So, it's better to speak of the free heap space here.

As for the artificial nature of edge cases, the code snippet above is certainly artificial, but there may be modules and user applications that face regressions introduced with this PR. I'm not 100% sure that there will any and I see the certain value introduced by this PR, so it would be nice to get more opinions here.

cc @nodejs/core

mmomtchev replied:
Proceeding by elimination:
There are only two ways of limiting the captured call stack before it is captured - one is by passing a reference of a function to V8 in `Error.captureStackTrace`. This is not usable without introducing some artificial function that will be able to find on the call stack - this is the old method. In theory, one could imagine some improved solution, where we never add more than once this function - the old method did it multiple times, creating closures at each call.

The other way is limiting the number of entries to a number known in advance - `Error.stackTraceLimit`. Either we estimate the maximum call stack that the Node internals are allowed to reach, either we try to deduce it in a dynamic way.

puzpuzpuz replied:
After some thinking, I would personally prefer option 2, i.e. the original solution, but with a modified "padding" (smth like 32 or even 64). Yes, it may lead to trimmed call stacks in produced errors if the padding is not enough, but at least it won't create unpredictable memory pressure.

puzpuzpuz commented on lib/internal/errors.js:
```diff
@@ -64,6 +65,16 @@ const prepareStackTrace = (globalThis, error, trace) => {
     return f(error, trace);
   }
 
+  for (let l = trace.length - 1; l >= 0; l--) {
```
nit: the old `hideStackFrames()` implementation had the following comment:
```
// Make sure the most outer `hideStackFrames()` function is used.
```

We should keep a similar comment around this loop to make the logic of iteration and further `splice()` more obvious.

mmomtchev replied:
Oh, yes, this was not intentional

puzpuzpuz commented on benchmark/misc/hidestackframes.js:
```diff
@@ -0,0 +1,50 @@
+
+const common = require('../common.js');
+
+/* there should be no significant performance differences
```
nit: `//` syntax is usually used for such comments. Also, the comment seems to be pretty obvious and could be omitted.

Trott commented:
Benchmark: https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/668/

puzpuzpuz commented:
Here is the benchmark output:
```
4:58:50                                                                 confidence improvement accuracy (*)    (**)   (***)
14:58:50  misc/hidestackframes.js n=100000 type='direct-call-noerr'                     -3.01 %       ±4.86%  ±6.48%  ±8.46%
14:58:50  misc/hidestackframes.js n=100000 type='direct-call-throw'             ***    -20.45 %       ±1.86%  ±2.49%  ±3.25%
14:58:50  misc/hidestackframes.js n=100000 type='hide-stackframes-noerr'        ***    137.14 %       ±8.16% ±10.86% ±14.16%
14:58:50  misc/hidestackframes.js n=100000 type='hide-stackframes-throw'        ***     -8.77 %       ±2.59%  ±3.44%  ±4.49%
14:58:50 
14:58:50 Be aware that when doing many comparisons the risk of a false-positive
14:58:50 result increases. In this case there are 4 comparisons, you can thus
14:58:50 expect the following amount of false-positive results:
14:58:50   0.20 false positives, when considering a   5% risk acceptance (*, **, ***),
14:58:50   0.04 false positives, when considering a   1% risk acceptance (**, ***),
14:58:50   0.00 false positives, when considering a 0.1% risk acceptance (***)
```

puzpuzpuz commented:
I also think, it's worth mentioning that this PR will close #35386. Does it make sense?

BridgeAR commented on lib/internal/errors.js:
```diff
@@ -64,6 +65,16 @@ const prepareStackTrace = (globalThis, error, trace) => {
 
+  for (let l = trace.length - 1; l >= 0; l--) {
+    const fn = trace[l].getFunctionName();
+    if (fn != null && fn.startsWith('__node_internal_hidden_')) {
```
This is currently run on all errors, not only on Node.js errors of which some stack frames should be hidden. As such, it's a general overhead that we should not introduce. Instead, let's verify that this code path is only triggered for Node.js errors.

mmomtchev replied:
This is true, I am at loss on how to do it without impacting the main path.

mmomtchev replied:
Maybe execute it for all errors triggered everywhere in Node.js - but not for user code - by detecting if it originated in Node.js - I think this is doable without changing ```hideStackFrames```

puzpuzpuz replied:
> This is currently run on all errors, not only on Node.js errors of which some stack frames should be hidden. As such, it's a general overhead that we should not introduce. Instead, let's verify that this code path is only triggered for Node.js errors.

It's a very good point. We should only enable the code path for built-in errors.

mmomtchev replied:
Done, the check is for captureLargerStackTrace as innermost stack element.
when processing an async exception, prepareStackTrace is called without the ```Error``` object, so there are not many other solutions

puzpuzpuz replied:
We have nested calls in some of validation functions, where only the outer function is wrapped with `hideStackFrames()`. For example, check `validateParameters` function from `lib/internal/crypto/hkdf.js`. If one of those nested validation functions (say, `validateString`) throws, the check here:
https://github.com/nodejs/node/pull/35644/commits/c205a80e9194153f3bb4203a21ef5b3f61c8df49#diff-670bf55805b781d9a3579f8bca9104c04d94af87cc33220149fd7d37b095ca1cR70
won't hold anymore. So, the user will get non-sanitized stack trace in the error.

I'd say, that the current approach to avoid the loop for user errors seems to be fragile and error-prone.

mmomtchev replied:
If processing a Node error with internals to hide, that last function on the stack will always be `captureLargerStackFrame`

puzpuzpuz replied:
> If processing a Node error with internals to hide, that last function on the stack will always be `captureLargerStackFrame`

Yes, you're right. It may be a good idea to leave a comment around the check and mention that the check will hold for built-in errors.

BridgeAR commented on lib/internal/errors.js:
```diff
@@ -286,45 +307,9 @@ function makeNodeErrorWithCode(Base, key) {
-    delete err.name;
-  }
+  const hidden = '__node_internal_hidden_' + fn.name;
+  ObjectDefineProperty(fn, 'name', { value: hidden });
```
These names are going to show up during profiling and likely while using a debugger. Changing the function names might therefore confuse people.

mmomtchev replied:
You still get the original name, but with a prefix (maybe I should reduce the prefix length) and besides, before, you were getting one ```hidden``` call every two frames - I think that if it is not better, it is at least not worse than before

BridgeAR replied:
My main pain point is that the name does not correspond with the actual function name anymore.

puzpuzpuz replied:
Yes, that's the main disadvantage of the approach.

mmomtchev replied:
```CallSite.getFunction()``` would have allowed for a much cleaner approach. Alas, it is getting removed in the next V8 version. And it was already gimped in strict mode.
<https://www.chromestatus.com/feature/5975936472186880>

BridgeAR reviewed (commented):
I am not really convinced that this implementation is a good idea.

mmomtchev commented:
> I also think, it's worth mentioning that this PR will close #35386. Does it make sense?

Where should I reference it? I did it in the commit message? Sorry, I am very new in this repo.

puzpuzpuz commented:
> Where should I reference it? I did it in the commit message? Sorry, I am very new in this repo.

Sorry, I didn't check the commit message. It's OK to reference the issue there.

aduh95 commented on lib/internal/errors.js:
```diff
@@ -460,37 +460,37 @@ function uvException(ctx) {
-    ex.port = port;
-  }
+    // Reducing the limit improves the performance significantly. We do not
+    // loose the stack frames due to the `captureStackTrace()` function that
```
```suggestion
    // lose the stack frames due to the `captureStackTrace()` function that
```

aduh95 commented on lib/internal/errors.js:
```diff
@@ -532,51 +535,50 @@ function errnoException(err, syscall, original) {
-    ex.port = port;
-  }
+    // Reducing the limit improves the performance significantly. We do not
+    // loose the stack frames due to the `captureStackTrace()` function that
```
```suggestion
    // lose the stack frames due to the `captureStackTrace()` function that
```

aduh95 reviewed (commented):
This needs a rebase.

mmomtchev commented on lib/internal/errors.js:
```diff
@@ -64,6 +66,19 @@ const prepareStackTrace = (globalThis, error, trace) => {
+        trace.splice(0, l + 1);
+        break;
+      }
+    }
```
That value is always updated before calling `new Error()` from Node. And when processing an exception originating inside user code, that part of the function does not run.
`Error.stackTraceLimit` is modified to be `Infinity` when processing Node exceptions - so that the full stack can be examined - then removing Node internals and splicing it to the correct size, before restoring the user value at the very end of `captureLargerStackTrace`

aduh95 replied:
OK, could you add a comment explaining that?

mmomtchev replied:
That part of the code has been changed, what is this referring to?

aduh95 replied:
I think it could be useful to add a comment explaining why you are using a cached value for `userStackTraceLimit` (what you explained in https://github.com/nodejs/node/pull/35644/files#r519655249 basically).

cclauss commented:
Please fix merge conflicts.

mcollina reviewed (approved):
lgtm

aduh95 commented on benchmark/misc/hidestackframes.js:
```diff
@@ -0,0 +1,46 @@
+  for (let i = 0; i < n; i++) {
+    try {
+      fn(value);
+    // eslint-disable-next-line no-unused-vars
+    } catch (e) {
```
```suggestion
    } catch {
```

aduh95 replied:
@mmomtchev Any reason to keep the unused variable here?

mmomtchev commented:
@benjamingr , since your PR, `hideStackFrames` should return a constructor - which cannot happen if the passed function is an arrow function. It worked before since `hideStackFrames` generated a new function, but not it is simply a rename
Since you are the only one to use `throw new` I guess it was an oversight?
Should I change this back to a normal call? In this PR?

mmomtchev commented:
@benjamingr I rebased and modified your code in `lib/internal/fs/promises.js` - replacing all `throw new lazyDOMException` with `throw lazyDOMException` - when `hideStackFrames` is used with an arrow function, it does not return a constructor anymore, it returns the same arrow function but renamed

benjamingr commented:
@mmomtchev actually consensus at the meeting was not to throw a `DOMException` because of compatibility concerns and throw an `AbortError` like https://github.com/nodejs/node/pull/36048/files#diff-cf533721360d04a028e31659e9b6cde7e4bd871200d13ce37d4a358be65bfd43R178 

So if you replace anything I would replace it with just `throw new AbortError` and import Error from errors/internal.

mmomtchev commented:
@benjamingr I prefer to avoid modifying the behavior of unrelated code in this PR, replacing `throw new` with simply `throw` is a simple syntax change
I hope I am not creating conflicts for your PR

benjamingr commented:
I haven't made a PR yet or wrote the code - the change is fine - it'll also change again anyway :]

Trott commented:
Changes that might affect benchmark results have been added to this PR since the benchmark CI run, so here's another benchmark CI run: https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/728/

aduh95 commented:
Benchmark results:

```
                                                               confidence improvement accuracy (*)    (**)   (***)
misc/hidestackframes.js n=100000 type='direct-call-noerr'                     -1.10 %       ±9.42% ±12.63% ±16.65%
misc/hidestackframes.js n=100000 type='direct-call-throw'             ***    -19.60 %       ±2.24%  ±2.99%  ±3.92%
misc/hidestackframes.js n=100000 type='hide-stackframes-noerr'        ***    176.09 %      ±25.35% ±34.14% ±45.27%
misc/hidestackframes.js n=100000 type='hide-stackframes-throw'        ***    -10.55 %       ±2.17%  ±2.89%  ±3.76%
```

mmomtchev commented:
@aduh95 @Trott the failed job failed with
```
13:09:51 collect2: fatal error: ld terminated with signal 9 [Killed]
```

mmomtchev commented:
@aduh95 is it normal for these tests to fail that much? The failing step is the linking of the debug version with all the symbols. Could be a memory limit problem maybe? Do the tests fail all the time?
One should be wary, because this PR modifies the way stack traces are generated, so we should be extra careful, but I don't think this is related?

Trott commented:
> One should be wary, because this PR modifies the way stack traces are generated, so we should be extra careful, but I don't think this is related?

I've seen that issue elsewhere in CI runs and I agree that it doesn't seem related.

Trott commented:
Would be nice to get one more review from someone, although it's not required. Maybe one of the commenters? @puzpuzpuz @aduh95 @BridgeAR

puzpuzpuz commented:
> Would be nice to get one more review from someone, although it's not required. Maybe one of the commenters? @puzpuzpuz @aduh95 @BridgeAR

@Trott I'm fine with the PR except for the concern described in https://github.com/nodejs/node/pull/35644#discussion_r505670254. But I certainly value the performance improvement on the success-path and don't want to block it and if there is a quorum, I'll be +1 on this PR.

aduh95 commented on lib/internal/errors.js:
```diff
@@ -75,6 +75,8 @@ const kTypes = [
 const MainContextError = Error;
 const overrideStackTrace = new SafeWeakMap();
 const kNoOverride = Symbol('kNoOverride');
+let userStackTraceLimit = Error.stackTraceLimit;
```
This assignment is not useful, is it?
```suggestion
let userStackTraceLimit;
```

mmomtchev replied:
You are right, it is a leftover from the initial version in which a user exception that came before the first Node exception could choke on the `undefined` value - it is impossible with the current code - but it doesn't cost anything and it could transform into a trap in the future

aduh95 replied:
It doesn't cost anything, but if you don't assign the value here, it makes it clearer that the code doesn't care of the value of `Error.stackTraceLimit` at this point. Also, if the code is changed in the future and introduced back the old behavior, I think it's helpful if it crashes to spot the mistake.

vdeturckheim reviewed (approved):
LGTM, thanks a lot to everyone who took part in this PR!

aduh95 commented:
@mmomtchev do you want to add comments in the code to resolve the last two open threads? If you prefer, I can add them myself by pushing commits to your branch. Other than that, I think this is ready to land.

aduh95 commented on lib/internal/errors.js:
```diff
@@ -84,6 +86,21 @@ const prepareStackTrace = (globalThis, error, trace) => {
+        break;
+      }
+    }
+    // userStackTraceLimit is the user value for Error.stackTraceLimit
+    // It is updated at every new exception
```
```suggestion
    // `userStackTraceLimit` is the user value for `Error.stackTraceLimit`,
    // it is updated at every new exception in `captureLargerStackTrace`.
```

aduh95 commented:
Landed in 656ce920a332

amcasey commented on lib/internal/errors.js:
```diff
@@ -413,6 +407,16 @@ function uvErrmapGet(name) {
+const captureLargerStackTrace = hideStackFrames(
+  function captureLargerStackTrace(err) {
+    userStackTraceLimit = Error.stackTraceLimit;
+    Error.stackTraceLimit = Infinity;
```
During module resolution, the TypeScript compiler has to probe many paths that turn out not to exist.  To mitigate the cost of creating so many `uvExceptions`, it sets `Error.stackTraceLimit` to zero (helpfully suggested [here](https://github.com/nodejs/node/pull/33716#pullrequestreview-424029712)).  From profiling, it looks like this change (specifically `captureLargerStackTrace node:internal/errors:450`) resulted in a pretty substantial perf regression.  Fortunately, newer version of TypeScript use the [new](https://github.com/nodejs/node/pull/33716/files) `throwIfNoEntry` flag, mitigating the issue, but older versions of TypeScript are likely to be substantially affected.  Have others reported performance issues in this area and, if so, would it be worthwhile to file a bug (I couldn't find an existing one)?

aduh95 replied:
You should be able to hack your way around it in Node.js v16.0.0+ by making the `stackTraceLimit` not writable:
```js
Reflect.defineProperty(Error, 'stackTraceLimit', {
  writable: false,
  value: 0,
});
```

amcasey replied:
I have to admit, that would not have occurred to me. :smile:

Fortunately, we already have a good mitigation for v16 - we just pass `throwIfNoEntry`.  The problem is more that code we've already shipped (i.e. TypeScript <= 4.1) lacks either mitigation and users will suddenly see a large slowdown.  In the repro I happened to notice this in (which, unfortunately, I can't share), the runtime jumps from 30 seconds to 48 seconds.

(If we end up filing a bug for this, I'll make sure to include a publicly-available sample.)

BridgeAR replied:
Ideally, we would have detected it's set to zero and not increased it in that case and only would have set it to it's current limit + a small number for the extra frames that should be removed.

BridgeAR commented on lib/internal/errors.js:
```diff
@@ -147,6 +162,27 @@ function lazyBuffer() {
 
+const addCodeToName = hideStackFrames(function addCodeToName(err, name, code) {
+  // Set the stack
+  err = captureLargerStackTrace(err);
```
This has caused some performance issues for all errors due to overcollecting stack frames where no frames were meant to be hidden.

BridgeAR commented:
With this approach it is actually not possible to limit the performance overhead to the methods that use `hideStackFrames()` anymore. We always either need to overallocate for all Node.js errors (which may cause a lot of overhead for deeply nested code) or we do not overallocate anything at all and the stack trace may stack empty.
